### PR TITLE
updated `cirq.Circuit` to `cirq.AbstractCircuit`

### PR DESCRIPTION
**Pull request by banrovegrie**
As requested in #4350.
Updated the argument of `program` from `cirq.Circuit` to `cirq.AbstractCircuit`. Changes are local to `sampler.py` only.

### CODE REVIEW

**Review by maffoo — APPROVED** [review]

**Review by vtomole — APPROVED** [review]

**Comment by mpharrigan** [comment]
@banrovegrie thanks for the contribution. There are some status checks that are failing

1. the format check can be resolved by running `check/format-incremental --apply` to automatically format your change
2. the type (mypy) check is a bit more sinister. Since we run type checking as part of the CI, your change has knock-on effects in places that implement the Sampler API. You'd likely need to update those type annotations as well

**Comment by banrovegrie** [comment]
@mpharrigan Thanks for describing the type check problem. I corrected the format issue and will try to get rid of the type errors one by one.

**Comment by mpharrigan** [comment]
you can run locally with `check/mypy`

**Comment by banrovegrie** [comment]
@mpharrigan Sorry to disturb you again. I fixed most of the type errors but now I am stuck while trying to fix the errors in other sampler files (total of 6). 

For example, in case of [pasqal_sampler.py](https://github.com/quantumlib/Cirq/blob/master/cirq-pasqal/cirq_pasqal/pasqal_sampler.py) changing type argument of program in `run_sweep` leads to error since `validate_circuit` can only take `program: cirq.Circuit` as its parameter.

Could you please guide me on how to fix this?

**Comment by mpharrigan** [comment]
There are ~three ways you can proceed. 

Just to be absolutely clear on the context here, we used to only have `Circuit` which is mutable. As such, a lot of the code just uses `Circuit` as its annotation whether or not it relies on mutation. When `FrozenCircuit` and `AbstractCircuit` were introduced, no one went through all the type annotations because as you're discovering there are lots of knock-on effects where you have to basically change _all_ the type annotations at once! And there are a lot that need changing!

**Option 1: the right way:** continue changing the type annotations to take `AbstractCircuit`. Like: change `Device.validate_circuit` to take `AbstractCircuit`. The type checker will make sure you're only using methods that both Circuit and FrozenCircuit have, so all the dependant functionality better not try to mutate the circuit. If it does: you either have to think whether you can fixup the functions to not rely on mutation or retreat and change all the dependant calls back to `Circuit` (please add a note why mutation is necessary, then)

**Option 2: the questionable shim way:** if you've had enough, update -- but do not close -- the original issue to say where you "left off". To keep the CI happy you can call `cast(Circuit, circuit)` to basically ignore the typecheck error. You should be pretty sure that all the calling code _isn't_ using mutation, or we could have unhappy users who thought it was ok to use FrozenCircuit/AbstractCircuit because of the annotation when in actuality it isn't

**Option 3: the runtime-penalty way:** If you've had enough, update -- but do not close -- the original issue to say where you "left off". To keep the CI happy, you can call `AbstractCircuit.unfreeze(copy=False)` which is a method on AbstractCircuit that turns either a Circuit or a FrozenCircuit into a Circuit. This is safe (unlike option 2) but we'll be (needlessly?) copying FrozenCircuits

**Comment by banrovegrie** [comment]
Update: Performed all the necessary changes by keeping on updating `Circuit` to `AbstractCircuit`, except in two specific situations.

1. In [single_qubit_readout_calibration.py](https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/experiments/single_qubit_readout_calibration.py), updated `results = sampler.run_batch(all_circuits, all_sweeps, repetitions=repetitions)` to the following.
```py
results = sampler.run_batch(cast(List[cirq.AbstractCircuit], all_circuits), all_sweeps, repetitions=repetitions)
```
2. In [cirq_rigetti/sampler.py](https://github.com/quantumlib/Cirq/blob/master/cirq-rigetti/cirq_rigetti/sampler.py) under `run_sweep` updated the parameter of `circuit` while returning `self.executer()`
```py
return self.executor(
      quantum_computer=self._quantum_computer,
      circuit=cast(cirq.Circuit, program.unfreeze(copy=False)),
      resolvers=resolvers,
      repetitions=repetitions,
      transformer=self.transformer,
)
```
Format and type checking performed locally before pushing the last commit.

**Comment by mpharrigan** [comment]
I don't think you need the `cast` if you're calling `unfreeze()`

**Comment by mpharrigan** [comment]
I'm confused why the cast is needed in  single_qubit_readout_calibration.py

Can it not figure out that `all_circuits` is a `List[cirq.Circuit]`? Maybe during the initial definition of `all_circuits` you can annotate it:

```python
all_circuits: List[cirq.Circuit] = [stuff]*123
```

**Comment by banrovegrie** [comment]
Yeah, I don't need `cast` after calling `unfreeze`, so changing that. 

Meanwhile, in `single_qubit_readout_calibration.py`, `all_circuits` is `List[cirq.Circuit]` by declaration and I am trying to cast that to `List[cirq.AbstractCircuit]` since in `run_batch()` I had previously type casted its first parameter i.e. a list of circuits to `List[cirq.AbstractCircuit]`.

**Comment by mpharrigan** [comment]
> Meanwhile, in single_qubit_readout_calibration.py, all_circuits is List[cirq.Circuit] by declaration and I am trying to cast that to List[cirq.AbstractCircuit] since in run_batch() I had previously type casted its first parameter i.e. a list of circuits to List[cirq.AbstractCircuit].

I want to understand this aspect, because I was under the strong impression that `List[cirq.Circuit]` would work when passing something accepting `List[cirq.AbstractCircuit]`. Is this not the case?

File: cirq-core/cirq/work/sampler.py [hunk 1]
```diff
@@ -197,7 +199,7 @@ async def run_sweep_async(
 
     def run_batch(
         self,
-        programs: List['cirq.Circuit'],
+        programs: List['cirq.AbstractCircuit'],
```
Comment (mpharrigan) [inline cirq-core/cirq/work/sampler.py:202]: I think this should be `Sequence['cirq.Abstractcircuit']`. Sequence is covariant whereas list is not

https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generics

**Comment by banrovegrie** [comment]
Yeah, my bad. I had to use `Sequence`. I no longer have to typecast anything.
Everything is fixed now.

**Comment by mpharrigan** [comment]
sweet! thanks for powering through

File: cirq-core/cirq/experiments/single_qubit_readout_calibration.py [hunk 2]
```diff
@@ -14,7 +14,7 @@
 """Single qubit readout experiments using parallel or isolated statistics."""
 import dataclasses
 import time
-from typing import Any, Dict, Iterable, List, Optional, TYPE_CHECKING
+from typing import Any, Dict, Iterable, List, Optional, TYPE_CHECKING, cast
```
Comment (mpharrigan) [inline cirq-core/cirq/experiments/single_qubit_readout_calibration.py:17]: unused import?

Comment (banrovegrie) [reply cirq-core/cirq/experiments/single_qubit_readout_calibration.py:17]: yeah i no longer need that updating

File: cirq-rigetti/cirq_rigetti/sampler.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 ##############################################################################
-from typing import List, Optional
+from typing import List, Optional, cast
```
Comment (mpharrigan) [inline cirq-rigetti/cirq_rigetti/sampler.py:16]: unused import?

Comment (banrovegrie) [reply cirq-rigetti/cirq_rigetti/sampler.py]: updated

**Review by mpharrigan — APPROVED** [review]

**Comment by banrovegrie** [comment]
@mpharrigan Thanks a lot for helping me out throughout.

**Pull request merged**